File: scripts/uber-scraper/uber-scraper.py
#!/usr/bin/env python3
"""
Uber Driver Earnings Scraper
Logs into drivers.uber.com and extracts weekly earnings summary.
Saves structured JSON to public/data/uber-earnings.json in Mission Control repo.

Usage:
    python3 uber-scraper.py              # scrape current week
    python3 uber-scraper.py --all        # scrape all available weeks
    python3 uber-scraper.py --dry-run    # scrape but don't save/commit

Requirements:
    pip install playwright python-dotenv
    playwright install chromium  (or use system chromium)
"""

import os
import sys
import json
import time
import argparse
import traceback
import subprocess
from datetime import datetime, timezone
from pathlib import Path

# Load .env
from dotenv import load_dotenv
load_dotenv(Path(__file__).parent / ".env")

# ── Config ────────────────────────────────────────────────────────────────────
CHROMIUM_PATH    = "/usr/local/bin/chromium"
REPO_ROOT        = Path(__file__).resolve().parents[2]   # mission-control/
OUTPUT_JSON      = REPO_ROOT / "public" / "data" / "uber-earnings.json"
AUTH_STATE_FILE  = Path(__file__).parent / "uber-auth-state.json"

UBER_EMAIL       = os.getenv("UBER_EMAIL", "")
UBER_PASSWORD    = os.getenv("UBER_PASSWORD", "")
DISCORD_WEBHOOK  = os.getenv("DISCORD_WEBHOOK_URL", "")
GITHUB_TOKEN     = os.getenv("GITHUB_TOKEN", "")

MAX_RETRIES      = 3
HEADLESS         = True

# ── Helpers ───────────────────────────────────────────────────────────────────
def log(msg):
    ts = datetime.now().strftime("%H:%M:%S")
    print(f"[{ts}] {msg}", flush=True)

def alert_discord(message: str):
    """Post failure alert to Discord webhook."""
    if not DISCORD_WEBHOOK:
        return
    try:
        import urllib.request, urllib.parse
        payload = json.dumps({"content": f"🚨 **Uber Scraper Alert**\n{message}"}).encode()
        req = urllib.request.Request(DISCORD_WEBHOOK, data=payload,
                                     headers={"Content-Type": "application/json"})
        urllib.request.urlopen(req, timeout=10)
    except Exception as e:
        log(f"  Discord alert failed: {e}")

def load_existing_earnings():
    if OUTPUT_JSON.exists():
        with open(OUTPUT_JSON) as f:
            return json.load(f)
    return {"weekly_summaries": [], "last_updated": None, "metadata": {}}

def save_earnings(data: dict, dry_run=False):
    data["last_updated"] = datetime.now(timezone.utc).isoformat()
    if dry_run:
        log("  [DRY RUN] Would save:")
        print(json.dumps(data, indent=2))
        return
    OUTPUT_JSON.parent.mkdir(parents=True, exist_ok=True)
    with open(OUTPUT_JSON, "w") as f:
        json.dump(data, f, indent=2)
    log(f"  Saved → {OUTPUT_JSON}")

def git_commit_push():
    """Auto-commit and push the updated JSON."""
    try:
        cwd = str(REPO_ROOT)
        subprocess.run(["git", "add", str(OUTPUT_JSON)], cwd=cwd, check=True)
        msg = f"auto: uber earnings update {datetime.now().strftime('%Y-%m-%d')}"
        subprocess.run(["git", "commit", "-m", msg], cwd=cwd, check=True)

        # Set remote URL with token if available
        if GITHUB_TOKEN:
            result = subprocess.run(["git", "remote", "get-url", "origin"],
                                    capture_output=True, text=True, cwd=cwd)
            remote_url = result.stdout.strip()
            if "https://" in remote_url and "@" not in remote_url:
                authed = remote_url.replace("https://", f"https://{GITHUB_TOKEN}@")
                subprocess.run(["git", "remote", "set-url", "origin", authed], cwd=cwd)

        subprocess.run(["git", "push", "origin", "main"], cwd=cwd, check=True)
        log("  ✅ Committed + pushed to GitHub")
    except subprocess.CalledProcessError as e:
        log(f"  ⚠️  Git push failed: {e}")


# ── Scraper ───────────────────────────────────────────────────────────────────
def parse_amount(text: str) -> float:
    """'$1,234.56' → 1234.56"""
    if not text:
        return 0.0
    cleaned = text.replace("$", "").replace(",", "").strip()
    try:
        return float(cleaned)
    except ValueError:
        return 0.0

def scrape_earnings(page, scrape_all=False) -> list:
    """Navigate to earnings page and extract weekly summaries."""
    log("  Navigating to earnings page…")
    page.goto("https://drivers.uber.com/earnings/activities", wait_until="networkidle", timeout=60000)
    page.wait_for_timeout(3000)

    # Check if we landed on a login page
    if "login" in page.url or "signin" in page.url:
        raise RuntimeError("Not authenticated — login redirect detected")

    log(f"  Page title: {page.title()}")
    log(f"  URL: {page.url}")

    # Take a debug screenshot
    screenshot_path = Path(__file__).parent / "debug-screenshot.png"
    page.screenshot(path=str(screenshot_path))
    log(f"  Screenshot saved: {screenshot_path}")

    weekly_summaries = []

    # ── Strategy 1: Look for weekly summary cards/rows ──────────────────────
    # Uber's earnings page shows pay periods — try multiple selectors
    selectors_to_try = [
        "[data-testid='weekly-summary']",
        "[data-testid='earnings-period']",
        ".earnings-summary",
        "[class*='earningsPeriod']",
        "[class*='weeklyEarning']",
        "[class*='PayStatement']",
        "//div[contains(text(), 'Week of')]",
        "//div[contains(text(), 'Pay period')]",
    ]

    found_element = None
    for sel in selectors_to_try:
        try:
            if sel.startswith("//"):
                elements = page.locator(f"xpath={sel}").all()
            else:
                elements = page.locator(sel).all()
            if elements:
                log(f"  Found {len(elements)} elements with selector: {sel}")
                found_element = sel
                break
        except Exception:
            continue

    # ── Strategy 2: Extract via page text + structured parsing ───────────────
    log("  Extracting page content for parsing…")
    content = page.content()

    # Try to find JSON data embedded in page (common in React apps)
    import re
    json_matches = re.findall(r'window\.__INITIAL_STATE__\s*=\s*({.+?});', content)
    if not json_matches:
        json_matches = re.findall(r'window\.__PRELOADED_STATE__\s*=\s*({.+?});', content)

    if json_matches:
        log("  Found embedded state JSON — extracting earnings…")
        try:
            state = json.loads(json_matches[0])
            # Try to find earnings in state tree
            earnings_data = find_earnings_in_state(state)
            if earnings_data:
                return earnings_data
        except Exception as e:
            log(f"  State parse failed: {e}")

    # ── Strategy 3: API intercept approach ───────────────────────────────────
    log("  Attempting API-based extraction…")
    api_data = extract_via_api(page)
    if api_data:
        return api_data

    # ── Strategy 4: DOM text scraping fallback ───────────────────────────────
    log("  Falling back to DOM text extraction…")
    return extract_from_dom(page)


def find_earnings_in_state(state: dict, depth=0) -> list:
    """Recursively search for earnings data in embedded state."""
    if depth > 8:
        return []
    if isinstance(state, dict):
        for key in ["earnings", "weeklyEarnings", "payStatements", "earningsSummary"]:
            if key in state and isinstance(state[key], (list, dict)):
                return normalize_earnings(state[key])
        for v in state.values():
            result = find_earnings_in_state(v, depth + 1)
            if result:
                return result
    elif isinstance(state, list):
        for item in state:
            result = find_earnings_in_state(item, depth + 1)
            if result:
                return result
    return []


def extract_via_api(page) -> list:
    """Intercept Uber's internal API calls for earnings data."""
    earnings_responses = []

    def handle_response(response):
        url = response.url
        if "earnings" in url.lower() and response.status == 200:
            try:
                body = response.json()
                earnings_responses.append(body)
                log(f"  Captured API response from: {url[:80]}")
            except Exception:
                pass

    page.on("response", handle_response)

    # Trigger page reload to capture API calls
    page.reload(wait_until="networkidle", timeout=60000)
    page.wait_for_timeout(5000)

    if earnings_responses:
        for resp in earnings_responses:
            normalized = normalize_earnings(resp)
            if normalized:
                return normalized

    return []


def extract_from_dom(page) -> list:
    """Last resort: parse visible text for dollar amounts and dates."""
    import re

    text = page.locator("body").inner_text()
    log(f"  Page text length: {len(text)} chars")

    # Look for weekly earnings patterns
    # "Apr 14 – Apr 20  $158.36  8 trips"
    pattern = re.compile(
        r'(Jan|Feb|Mar|Apr|May|Jun|Jul|Aug|Sep|Oct|Nov|Dec)\s+\d+\s*[–\-—]\s*'
        r'(Jan|Feb|Mar|Apr|May|Jun|Jul|Aug|Sep|Oct|Nov|Dec)\s+\d+.*?\$(\d+\.?\d*)',
        re.DOTALL
    )
    matches = pattern.findall(text)
    if matches:
        log(f"  Found {len(matches)} date-range matches")

    # Build minimal summary from what we can find
    return []


def normalize_earnings(raw) -> list:
    """Normalize API response into our standard weekly summary format."""
    summaries = []
    if isinstance(raw, dict):
        raw = [raw]
    if not isinstance(raw, list):
        return []
    for item in raw:
        if not isinstance(item, dict):
            continue
        summary = {
            "period_start": item.get("periodStart") or item.get("startDate") or item.get("weekStart", ""),
            "period_end":   item.get("periodEnd")   or item.get("endDate")   or item.get("weekEnd", ""),
            "total_earnings": float(item.get("totalEarnings") or item.get("earnings") or 0),
            "total_trips":    int(item.get("totalTrips")    or item.get("trips")   or 0),
            "breakdown": {
                "base_fare":   float(item.get("baseFare")   or 0),
                "surge":       float(item.get("surge")      or 0),
                "promotions":  float(item.get("promotions") or 0),
                "tips":        float(item.get("tips")       or item.get("tip") or 0),
                "expenses":    float(item.get("expenses")   or 0),
            },
            "scraped_at": datetime.now(timezone.utc).isoformat(),
        }
        if summary["period_start"] or summary["total_earnings"]:
            summaries.append(summary)
    return summaries


# ── Auth ──────────────────────────────────────────────────────────────────────
def login(page, context):
    """Log into Uber driver portal. Saves auth state for reuse."""
    log("  Starting login flow…")
    page.goto("https://auth.uber.com/v2/", wait_until="networkidle", timeout=60000)
    page.wait_for_timeout(2000)

    # If already logged in (saved state), skip
    if "drivers.uber.com" in page.url or page.url.startswith("https://drivers.uber.com"):
        log("  Already authenticated via saved state")
        return True

    # Enter email
    try:
        email_input = page.locator("input[type='email'], input[name='email'], #useridInput")
        email_input.first.fill(UBER_EMAIL)
        page.keyboard.press("Enter")
        page.wait_for_timeout(2000)
    except Exception as e:
        log(f"  Email input not found: {e}")
        return False

    # Enter password
    try:
        pwd_input = page.locator("input[type='password'], #password")
        pwd_input.first.fill(UBER_PASSWORD)
        page.keyboard.press("Enter")
        page.wait_for_timeout(3000)
    except Exception as e:
        log(f"  Password input not found: {e}")
        return False

    # Check for 2FA / phone verification
    if "verify" in page.url or "2fa" in page.url.lower():
        log("  ⚠️  2FA required — cannot proceed headlessly")
        log("  Run with HEADLESS=false to complete 2FA manually, then save auth state")
        return False

    # Check if login succeeded
    page.wait_for_timeout(3000)
    if "drivers.uber.com" in page.url:
        # Save auth state for future runs
        context.storage_state(path=str(AUTH_STATE_FILE))
        log(f"  Auth state saved → {AUTH_STATE_FILE}")
        return True

    log(f"  Login may have failed. Current URL: {page.url}")
    return False


# ── Main ──────────────────────────────────────────────────────────────────────
def run(scrape_all=False, dry_run=False, commit=True):
    from playwright.sync_api import sync_playwright

    if not UBER_EMAIL or not UBER_PASSWORD:
        log("❌ UBER_EMAIL and UBER_PASSWORD must be set in .env")
        sys.exit(1)

    last_error = None
    for attempt in range(1, MAX_RETRIES + 1):
        log(f"\n── Attempt {attempt}/{MAX_RETRIES} ─────────────────────────────")
        try:
            with sync_playwright() as p:
                # Use saved auth state if it exists
                context_opts = {
                    "executable_path": CHROMIUM_PATH if os.path.exists(CHROMIUM_PATH) else None,
                }
                if AUTH_STATE_FILE.exists():
                    log("  Loading saved auth state…")
                    context_opts["storage_state"] = str(AUTH_STATE_FILE)

                browser = p.chromium.launch(
                    headless=HEADLESS,
                    executable_path=CHROMIUM_PATH if os.path.exists(CHROMIUM_PATH) else None,
                    args=["--no-sandbox", "--disable-dev-shm-usage"],
                )
                context = browser.new_context(
                    user_agent="Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) "
                               "AppleWebKit/537.36 (KHTML, like Gecko) "
                               "Chrome/124.0.0.0 Safari/537.36",
                    storage_state=str(AUTH_STATE_FILE) if AUTH_STATE_FILE.exists() else None,
                    viewport={"width": 1280, "height": 900},
                )
                page = context.new_page()

                # Navigate to earnings page (redirect to login if not authed)
                page.goto("https://drivers.uber.com/earnings/activities",
                          wait_until="networkidle", timeout=60000)
                page.wait_for_timeout(2000)

                # If redirected to login, do the login flow
                if "login" in page.url or "auth.uber.com" in page.url or "signin" in page.url:
                    log("  Not authenticated — logging in…")
                    success = login(page, context)
                    if not success:
                        raise RuntimeError("Login failed")
                    # Navigate back to earnings
                    page.goto("https://drivers.uber.com/earnings/activities",
                              wait_until="networkidle", timeout=60000)
                    page.wait_for_timeout(3000)
                else:
                    log("  ✅ Already authenticated")

                # Scrape
                weekly = scrape_earnings(page, scrape_all=scrape_all)
                browser.close()

                if not weekly:
                    log("  ⚠️  No weekly data extracted — check debug-screenshot.png")
                    # Still save whatever we have (won't overwrite good data)
                    weekly = []

                # Merge with existing data
                existing = load_existing_earnings()
                existing = merge_earnings(existing, weekly)
                existing["metadata"] = {
                    "source": "drivers.uber.com",
                    "scrape_method": "playwright",
                    "last_scrape_attempt": datetime.now(timezone.utc).isoformat(),
                }
                save_earnings(existing, dry_run=dry_run)

                if not dry_run and commit and weekly:
                    git_commit_push()

                log(f"\n✅ Done — {len(weekly)} week(s) extracted")
                return weekly

        except Exception as e:
            last_error = e
            log(f"  ❌ Error: {e}")
            if attempt < MAX_RETRIES:
                wait = attempt * 15
                log(f"  Retrying in {wait}s…")
                time.sleep(wait)

    # All retries failed
    msg = f"Failed after {MAX_RETRIES} attempts: {last_error}"
    log(f"\n❌ {msg}")
    alert_discord(msg)
    sys.exit(1)


def merge_earnings(existing: dict, new_weeks: list) -> dict:
    """Merge new weekly summaries, deduplicating by period_start."""
    summaries = existing.get("weekly_summaries", [])
    existing_keys = {w.get("period_start") for w in summaries if w.get("period_start")}

    added = 0
    for week in new_weeks:
        key = week.get("period_start")
        if key and key not in existing_keys:
            summaries.append(week)
            existing_keys.add(key)
            added += 1
        elif not key:
            summaries.append(week)
            added += 1

    # Sort by period_start descending
    summaries.sort(key=lambda w: w.get("period_start", ""), reverse=True)
    existing["weekly_summaries"] = summaries
    log(f"  Merged: {added} new week(s) added ({len(summaries)} total)")
    return existing


# ── CLI ───────────────────────────────────────────────────────────────────────
if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Uber Earnings Scraper")
    parser.add_argument("--all",     action="store_true", help="Scrape all available weeks")
    parser.add_argument("--dry-run", action="store_true", help="Don't save or commit")
    parser.add_argument("--no-commit", action="store_true", help="Save but don't git push")
    parser.add_argument("--visible", action="store_true", help="Run with visible browser (debug)")
    args = parser.parse_args()

    if args.visible:
        HEADLESS = False

    log("🚗 Uber Earnings Scraper starting…")
    log(f"   Email: {UBER_EMAIL or '(not set)'}")
    log(f"   Output: {OUTPUT_JSON}")
    log(f"   Auth state: {'found' if AUTH_STATE_FILE.exists() else 'not found (will login)'}")

    run(
        scrape_all=args.all,
        dry_run=args.dry_run,
        commit=not args.no_commit,
    )

File: scripts/uber-scraper/uber-scraper.js
#!/usr/bin/env node
/**
 * Uber Driver Earnings Scraper
 * 
 * Logs into Uber driver dashboard → extracts weekly earnings summary
 * → saves to Mission Control public/data/uber-earnings.json
 * 
 * Auth strategy: Persistent browser context (auth-state.json)
 * - First run: interactive login (prompts for OTP if needed)
 * - Subsequent runs: reuses saved session (fully headless)
 * 
 * Usage:
 *   node uber-scraper.js              # Normal run
 *   node uber-scraper.js --login      # Force fresh login
 *   node uber-scraper.js --dry-run    # Extract only, no file write
 *   node uber-scraper.js --weeks 4    # Pull last 4 weeks (default: 2)
 */

const { chromium } = require('./node_modules/playwright');
const fs = require('fs');
const path = require('path');
const readline = require('readline');

// ─── Config ───────────────────────────────────────────────────────────────────

const UBER_EMAIL    = process.env.UBER_EMAIL    || 'ericmills71@gmail.com';
const UBER_PHONE    = process.env.UBER_PHONE    || '';
const UBER_PASSWORD = process.env.UBER_PASSWORD || 'Wicked71!';

const EARNINGS_URL  = 'https://drivers.uber.com/earnings/activities';
const LOGIN_URL     = 'https://auth.uber.com/login';

const SCRIPT_DIR    = __dirname;
const AUTH_FILE     = path.join(SCRIPT_DIR, 'auth-state.json');
const OUTPUT_FILE   = path.join(
  '/home/ubuntu/wlp/projects/mission-control/public/data',
  'uber-earnings.json'
);
const LOG_FILE      = path.join(SCRIPT_DIR, 'cron.log');

const ARGS          = process.argv.slice(2);
const FORCE_LOGIN   = ARGS.includes('--login');
const DRY_RUN       = ARGS.includes('--dry-run');
const WEEKS_IDX     = ARGS.indexOf('--weeks');
const WEEKS_TO_PULL = WEEKS_IDX >= 0 ? parseInt(ARGS[WEEKS_IDX + 1]) || 2 : 2;

// ─── Logging ──────────────────────────────────────────────────────────────────

function log(msg) {
  const ts = new Date().toISOString();
  const line = `[${ts}] ${msg}`;
  console.log(line);
  try { fs.appendFileSync(LOG_FILE, line + '\n'); } catch {}
}

function logError(msg, err) {
  log(`❌ ${msg}: ${err?.message || err}`);
}

// ─── OTP Prompt (for interactive first login) ─────────────────────────────────

function promptOTP(prompt) {
  return new Promise(resolve => {
    const rl = readline.createInterface({ input: process.stdin, output: process.stdout });
    rl.question(prompt, answer => { rl.close(); resolve(answer.trim()); });
  });
}

// ─── Login Flow ───────────────────────────────────────────────────────────────

async function doLogin(page) {
  log('🔓 Starting Uber login flow...');

  await page.goto(LOGIN_URL, { waitUntil: 'domcontentloaded', timeout: 30000 });
  await page.waitForTimeout(2000);

  const url = page.url();
  log(`📄 After goto: ${url}`);

  // Fill identifier (email or phone)
  const identifier = UBER_EMAIL || UBER_PHONE;
  try {
    // Uber uses a single input for email/phone on first step
    await page.waitForSelector('input[type="text"], input[type="email"], input[id*="USER_ID"]', { timeout: 10000 });
    await page.fill('input[type="text"], input[type="email"], input[id*="USER_ID"]', identifier);
    await page.waitForTimeout(500);

    // Click "Next" / "Continue"
    const nextBtn = page.locator('button[type="submit"], button:has-text("Next"), button:has-text("Continue")').first();
    await nextBtn.click();
    await page.waitForTimeout(2000);
  } catch (e) {
    log(`⚠️  Identifier step issue: ${e.message}`);
  }

  // Password step
  try {
    const pwInput = await page.waitForSelector('input[type="password"]', { timeout: 8000 });
    if (pwInput) {
      await page.fill('input[type="password"]', UBER_PASSWORD);
      await page.waitForTimeout(500);
      const submitBtn = page.locator('button[type="submit"]').first();
      await submitBtn.click();
      await page.waitForTimeout(3000);
    }
  } catch (e) {
    log(`⚠️  Password step: ${e.message} — may have skipped to OTP`);
  }

  // OTP step — Uber often sends SMS or email verification
  const currentUrl = page.url();
  if (currentUrl.includes('verify') || currentUrl.includes('otp') || currentUrl.includes('mfa')) {
    log('📱 OTP/verification required — check your phone or email');
    const otp = await promptOTP('Enter OTP code: ');
    try {
      await page.fill('input[type="text"], input[type="number"], input[autocomplete="one-time-code"]', otp);
      await page.waitForTimeout(500);
      const submitBtn = page.locator('button[type="submit"]').first();
      await submitBtn.click();
      await page.waitForTimeout(3000);
    } catch (e) {
      log(`⚠️  OTP entry failed: ${e.message}`);
    }
  }

  // Verify we're logged in
  const finalUrl = page.url();
  log(`📄 Post-login URL: ${finalUrl}`);

  if (finalUrl.includes('drivers.uber.com') || finalUrl.includes('earnings')) {
    log('✅ Login successful');
    return true;
  }

  log('⚠️  Login may have failed — unexpected URL after auth');
  return false;
}

// ─── Earnings Extraction ──────────────────────────────────────────────────────

async function extractEarnings(page) {
  log(`🔍 Navigating to earnings: ${EARNINGS_URL}`);

  await page.goto(EARNINGS_URL, { waitUntil: 'domcontentloaded', timeout: 30000 });
  await page.waitForTimeout(4000);

  const url = page.url();
  log(`📄 Earnings page URL: ${url}`);

  if (url.includes('login') || url.includes('auth.uber')) {
    throw new Error('Redirected to login — session expired');
  }

  // ── Strategy 1: Intercept XHR/Fetch API responses ─────────────────────────
  log('🔍 Attempting API intercept strategy...');

  let apiData = null;
  page.on('response', async response => {
    const respUrl = response.url();
    if (respUrl.includes('/api/query') || respUrl.includes('/earnings') || respUrl.includes('/driver-profiles')) {
      try {
        const json = await response.json();
        if (json?.data?.driverEarnings || json?.earnings || json?.weeklyEarnings) {
          apiData = json;
          log(`✅ API intercept hit: ${respUrl}`);
        }
      } catch {}
    }
  });

  // Reload to trigger API calls
  await page.reload({ waitUntil: 'networkidle', timeout: 30000 });
  await page.waitForTimeout(3000);

  if (apiData) {
    log('✅ Got earnings data via API intercept');
    return parseApiData(apiData);
  }

  // ── Strategy 2: __NEXT_DATA__ / window state ───────────────────────────────
  log('🔍 Trying __NEXT_DATA__ extraction...');
  const nextData = await page.evaluate(() => {
    const el = document.getElementById('__NEXT_DATA__');
    if (el) {
      try { return JSON.parse(el.textContent); } catch {}
    }
    // Also check window.__PRELOADED_STATE__ or similar
    return window.__PRELOADED_STATE__ || window.__INITIAL_STATE__ || null;
  });

  if (nextData) {
    const parsed = parseNextData(nextData);
    if (parsed) {
      log('✅ Got earnings via __NEXT_DATA__');
      return parsed;
    }
  }

  // ── Strategy 3: DOM scraping ───────────────────────────────────────────────
  log('🔍 Falling back to DOM scraping...');
  const domData = await scrapeDOM(page);
  if (domData) {
    log('✅ Got earnings via DOM scrape');
    return domData;
  }

  // ── Strategy 4: Take screenshot for debugging ─────────────────────────────
  const screenshotPath = path.join(SCRIPT_DIR, 'debug-screenshot.png');
  await page.screenshot({ path: screenshotPath, fullPage: true });
  log(`📸 Screenshot saved: ${screenshotPath}`);

  throw new Error('Could not extract earnings data — screenshot saved for debugging');
}

// ─── Parse Helpers ────────────────────────────────────────────────────────────

function parseApiData(json) {
  // Uber's internal API varies — try common shapes
  const raw = json?.data?.driverEarnings?.weeklyEarnings
    || json?.weeklyEarnings
    || json?.earnings?.weeks
    || [];

  return raw.map(week => buildWeekEntry(week));
}

function parseNextData(data) {
  try {
    const props = data?.props?.pageProps;
    const earnings = props?.initialState?.earnings
      || props?.earnings
      || props?.weeklyEarnings;
    if (!earnings) return null;
    return Array.isArray(earnings) ? earnings.map(buildWeekEntry) : null;
  } catch {
    return null;
  }
}

async function scrapeDOM(page) {
  return await page.evaluate(() => {
    const results = [];

    // Look for weekly summary cards
    const cards = document.querySelectorAll(
      '[data-testid*="week"], [class*="WeekCard"], [class*="weekly-summary"], [class*="earnings-row"]'
    );

    cards.forEach(card => {
      const text = card.innerText;
      const amountMatch = text.match(/\$[\d,]+\.?\d*/g);
      const dateMatch = text.match(/(\w{3}\s+\d{1,2})\s*[-–]\s*(\w{3}\s+\d{1,2})/);
      const tripsMatch = text.match(/(\d+)\s+trip/i);

      if (amountMatch || dateMatch) {
        results.push({
          raw: text.substring(0, 300),
          amounts: amountMatch,
          dateRange: dateMatch ? dateMatch[0] : null,
          trips: tripsMatch ? parseInt(tripsMatch[1]) : null,
        });
      }
    });

    // Also grab overall summary section
    const summaryEl = document.querySelector(
      '[class*="summary"], [class*="total-earnings"], [data-testid*="summary"]'
    );
    const summaryText = summaryEl?.innerText || '';

    return results.length > 0 ? results : (summaryText ? [{ raw: summaryText }] : null);
  });
}

function buildWeekEntry(week) {
  // Normalize various API shapes into our standard format
  const startDate = week.startDate || week.start_date || week.periodStart || '';
  const endDate   = week.endDate   || week.end_date   || week.periodEnd   || '';

  return {
    period_start: startDate,
    period_end:   endDate,
    total_earnings:  parseFloat(week.totalEarnings  || week.total_earnings  || week.gross || 0),
    total_trips:     parseInt(  week.totalTrips     || week.total_trips     || week.trips || 0),
    breakdown: {
      base_fare:   parseFloat(week.baseFare    || week.base_fare   || 0),
      surge:       parseFloat(week.surge       || week.surgePay    || 0),
      tips:        parseFloat(week.tips        || week.tip         || 0),
      promotions:  parseFloat(week.promotions  || week.boosts      || 0),
      expenses:    parseFloat(week.expenses    || week.deductions  || 0),
    },
    net_payout: parseFloat(week.netPayout || week.net_payout || week.payout || 0),
    scraped_at: new Date().toISOString(),
  };
}

// ─── Load / Merge / Save JSON ─────────────────────────────────────────────────

function loadExisting() {
  if (fs.existsSync(OUTPUT_FILE)) {
    try { return JSON.parse(fs.readFileSync(OUTPUT_FILE, 'utf8')); } catch {}
  }
  return { weekly_summaries: [], last_updated: null, scraper_version: '1.0' };
}

function mergeWeeks(existing, fresh) {
  const byKey = {};

  // Index existing entries
  for (const w of (existing.weekly_summaries || [])) {
    const key = `${w.period_start}:${w.period_end}`;
    byKey[key] = w;
  }

  // Merge/overwrite with fresh
  for (const w of fresh) {
    const key = `${w.period_start}:${w.period_end}`;
    byKey[key] = w;
  }

  // Sort newest first
  return Object.values(byKey).sort((a, b) =>
    (b.period_start || '').localeCompare(a.period_start || '')
  );
}

function saveData(weeks) {
  const data = loadExisting();
  data.weekly_summaries = mergeWeeks(data, weeks);
  data.last_updated = new Date().toISOString();

  if (DRY_RUN) {
    log('🧪 DRY RUN — data extracted but not written:');
    console.log(JSON.stringify(data, null, 2));
    return;
  }

  fs.mkdirSync(path.dirname(OUTPUT_FILE), { recursive: true });
  fs.writeFileSync(OUTPUT_FILE, JSON.stringify(data, null, 2));
  log(`✅ Saved ${weeks.length} week(s) to ${OUTPUT_FILE}`);
}

// ─── Git Commit + Push ────────────────────────────────────────────────────────

function gitCommitPush() {
  const { execSync } = require('child_process');
  const repoDir = '/home/ubuntu/wlp/projects/mission-control';
  try {
    execSync(`cd "${repoDir}" && git add public/data/uber-earnings.json && git commit -m "chore: update uber earnings [$(date +%Y-%m-%d)]" && git push`, { stdio: 'inherit' });
    log('✅ Git commit + push complete');
  } catch (e) {
    logError('Git commit failed', e);
  }
}

// ─── Main ─────────────────────────────────────────────────────────────────────

async function main() {
  log('═══════════════════════════════════════════════');
  log('🚀 Uber Earnings Scraper starting');
  log(`   Weeks to pull: ${WEEKS_TO_PULL} | Dry run: ${DRY_RUN} | Force login: ${FORCE_LOGIN}`);

  const hasAuthState = fs.existsSync(AUTH_FILE) && !FORCE_LOGIN;
  const headless     = hasAuthState; // Interactive if no saved session

  if (!headless) {
    log('⚠️  No saved auth state — running in headed mode for interactive login');
    log('   (Set DISPLAY or run on a machine with a screen, or use --login on Mac)');
  }

  let browser, context;

  for (let attempt = 1; attempt <= 3; attempt++) {
    try {
      log(`🔄 Attempt ${attempt}/3`);

      browser = await chromium.launch({
        headless,
        args: ['--no-sandbox', '--disable-setuid-sandbox', '--disable-dev-shm-usage'],
      });

      const contextOptions = hasAuthState
        ? { storageState: AUTH_FILE }
        : {};

      context = await browser.newContext({
        ...contextOptions,
        userAgent: 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/122.0.0.0 Safari/537.36',
        viewport: { width: 1280, height: 800 },
      });

      const page = await context.newPage();

      // If no auth state, do login
      if (!hasAuthState) {
        const loginOk = await doLogin(page);
        if (!loginOk) throw new Error('Login failed');

        // Save auth state for future headless runs
        await context.storageState({ path: AUTH_FILE });
        log(`💾 Auth state saved to ${AUTH_FILE}`);
      }

      // Extract earnings
      const weeks = await extractEarnings(page);

      if (!weeks || weeks.length === 0) {
        throw new Error('No earnings data extracted');
      }

      log(`📊 Extracted ${weeks.length} week(s) of earnings`);

      // Save + push
      saveData(weeks);
      if (!DRY_RUN) gitCommitPush();

      await browser.close();
      log('✅ Scraper finished successfully');
      process.exit(0);

    } catch (err) {
      logError(`Attempt ${attempt} failed`, err);

      // If session expired, clear auth state and retry with login
      if (err.message.includes('session expired') || err.message.includes('login')) {
        log('🗑️  Clearing auth state — session expired');
        try { fs.unlinkSync(AUTH_FILE); } catch {}
      }

      if (browser) await browser.close().catch(() => {});

      if (attempt < 3) {
        log(`⏳ Waiting 10s before retry...`);
        await new Promise(r => setTimeout(r, 10000));
      }
    }
  }

  log('❌ All attempts failed — exiting with error');
  process.exit(1);
}

main();
